refactor codebase to use name buildspec

shahzebsiddiqui opened the pull request:
This PR will clean up a few things that were bit of a mess.
- First attempt to refactor code base to use the word ``Buildspec``. For variables we refer to as ``buildspec``. 
- rename option``buildtest build --config`` to ``buildtest build --buildspec`` 
- Rename schema ``outer.schema.json`` --> ``global.schema.json``
- Rename file ``default.yml`` --> ``settings.yml``
- we remove ``init_log`` not needed since it was used once. 
- Add regtest for get.py for clone feature.
- change behavior of is_dir and is_file to return a boolean and these methods run ``resolve_path`` first
- Remove ``create_file`` since its not used in codebase we can add this in if it's needed .
- ``create_dir`` won't use ``resolve_path`` since input directory is non-existent so we run the resolve manually.
- Reuse ``is_file`` and ``is_dir`` in regtest in places of os.path.exists when applicable to test buildtest modules. 
- refactor a few assert statements to be condensed such as ``assert True is statement`` to ``assert statement``

shahzebsiddiqui committed:
add regtest for buildtest get feature for cloning tutorials repo.
remote init_log method since it was used once in the clone feature so
that is now using system logger.
Remove unused method create_file since it was only referenced in regtest
and not actual code base.
Refactor tests for file operation.
Change behavior of is_dir and is_file to return a boolean.
Also create_dir, is_file and is_dir will call resolve_path first before
checking for existence.
Change behavior or resolve_path to return file no matter what, it should
be checked with is_file and is_dir if file exists.

shahzebsiddiqui committed:
fix bug where after exclusion there could be no files left so we need
to exit again. We exit for first time if there are no files discovered
because we want to terminate early without going into exclude operation
with an empty list.
Also we moved the output of discovered configs after we run the
exclusion to ensure we print the files that will be processed by
buildtest

shahzebsiddiqui committed:
fix black check

Signed-off-by: Shahzeb Siddiqui <shahzebmsiddiqui@gmail.com>

vsoch commented on buildtest/menu/build.py:
```diff
@@ -166,6 +162,15 @@ def func_build_subcmd(args):
         logger.debug(f"Configuration List after applying exclusion: {config_files}")
 
+        # if no files remain after exclusion let's stop now.
+        if not config_files:
```
Just to check here - you don't want any message printed to the log, but you are passing to sys.exit?

shahzebsiddiqui replied:
yes that's right, buildtest is there to inform user after discovery or after exclusion step if there are no configs. I know it's repeated twice after each stage. Atleast user should know what buildtest is trying to do.

vsoch commented on buildtest/utils/file.py:
```diff
@@ -11,60 +11,62 @@
-    :raises BuildTestError: sub-class of Exception
     :return: returns True if file exists otherwise terminates with an exception
-    :rtype: True if successful, otherwise throws an exception
+    :rtype: bool
```
Should this be rtype (does it mean return type or is it a typo?)

shahzebsiddiqui replied:
Yea this is return type, there are built-in types atleast documented in https://sphinx-rtd-tutorial.readthedocs.io/en/latest/docstrings.html i dont think it's an official standard from docstring but what happens is this formatting helps document all variables and return type that makes API docs look much better. 

Some IDEs have built-in support to auto-fill some sections atleast PyCharm can do this. https://www.jetbrains.com/help/pycharm/using-docstrings-to-specify-types.html

vsoch commented on buildtest/utils/file.py:
```diff
@@ -125,18 +106,19 @@ def resolve_path(path):
     real_path = os.path.realpath(path)
-    if os.path.exists(real_path):
-        return real_path
+    # if os.path.exists(real_path):
```
Why is this commented out now?

shahzebsiddiqui replied:
I will revert this back, it was addressing an issue in ``is_file`` and ``is_dir`` but i have a better solution while preserving this. Stay tuned.

shahzebsiddiqui replied:
I have reverted this behavior now in commit https://github.com/HPC-buildtest/buildtest-framework/pull/270/commits/2a8d6749580f2abad0c3e4c65a85b5a3d9e580e3 .

vsoch commented on tests/menu/test_get.py:
```diff
@@ -0,0 +1,27 @@
+from buildtest.utils.file import is_dir, create_dir
+
+
+def test_clone():
```
This should use pytest tmp_path.

shahzebsiddiqui replied:
Okay this is now fixed in https://github.com/HPC-buildtest/buildtest-framework/pull/270/commits/542e243641b7a88a091dac4fbaf5869fe99ccfa9

vsoch commented on tests/utils/test_file.py:
```diff
@@ -1,84 +1,53 @@
+    assert not is_file(file_name)
 
+    file1 = "~/.profile"
+    file2 = "$HOME/.profile"
```
Can we be sure these are universally present?

shahzebsiddiqui replied:
I can't say definitely, but knowing as sysadmin and the Xenial image provides a basic setup but it's not documented about startup configs travis https://docs.travis-ci.com/user/reference/xenial/. I'd say travis user is probably going to be bash, I have not tested this but could make some changes to ``.travis.yml`` for confirmation. 

Since test were passing it would mean these files exist.

vsoch replied:
Did you finish this comment?

shahzebsiddiqui replied:
That was typo in my last comment. Anyhow i didn't touch this regtest. The test will pass on most linux distros if it's a concern we can refactor test if needed. I needed something to validate with user and shell expansion. For user expansion it has to be a file that must exist and startup files are usually found as part of user creation (``useradd``)

vsoch commented on tests/utils/test_file.py:
```diff
@@ -1,84 +1,53 @@
 
+def test_directory_expansion():
+    dir1 = "$HOME"
+    dir2 = "~"
```
Same question as above.

vsoch commented on tests/utils/test_file.py:
```diff
@@ -1,84 +1,53 @@
-    assert True is os.path.isdir(os.path.expanduser("~/x/y/z"))
+    dir1 = "$HOME/a/b/c"
+    dir2 = "~/x/y/z"
+    create_dir(dir1)
```
Same question as above. It's generally good practice to use os.path.sep instead of hard coding like that.

shahzebsiddiqui replied:
i used ``os.path.join`` i guess the purpose of these two tests was to check shell and user expansion
with recursive directory creation. Change be seen in https://github.com/HPC-buildtest/buildtest-framework/pull/270/commits/2a8d6749580f2abad0c3e4c65a85b5a3d9e580e3#diff-a42582ff8b56b5783fdd220f2eb5ffab

shahzebsiddiqui committed:
Refactor majority of code-base to use buildspec instead of config,
rename methods, variables, update docstring, output and logging to
have correct usage.
Refactor some regression test to account for method name changes.
Revert behavior of resolve_conf to return None when file doesn't exist.
In create_dir method we can't invoke resolve_path method but rather
we run the resolve the paths manually.

change option --config to --buildspec with short option -bs
Rename outer.schema.json to global.schema.json
Rename file default.yml to settings.yml

vsoch commented on buildtest/buildsystem/base.py:
```diff
@@ -69,28 +69,28 @@ def __init__(self, config_file):
-        if not os.path.exists(self.config_file):
-            sys.exit("Build configuration %s does not exist." % self.config_file)
+        if not os.path.exists(self.buildspec_file):
+            sys.exit("BuildSpec File: %s does not exist." % self.buildspec_file)
```
Remember the capitalization stuffs we discussed!

shahzebsiddiqui replied:
oh that is a typo in line ``sys.exit("BuildSpec File: %s does not exist`` This should be ``Buildspec`` the S was caps by accident. 

Yea for variables it was assumed to be all lower case. I can fix this

vsoch commented on buildtest/buildsystem/base.py:
```diff
@@ -69,28 +69,28 @@ def __init__(self, config_file):
 
-        elif os.path.isdir(self.config_file):
+        elif os.path.isdir(self.buildspec_file):
             sys.exit(
```
For a lot of these saying “Buildspec file” is redundant, a buildspec is a file.

shahzebsiddiqui replied:
I will fix this and change to ``self.buildspec``

vsoch commented on buildtest/buildsystem/base.py:
```diff
@@ -105,24 +105,24 @@ def __repr__(self):
            to buildtest. If a version is provided, honor it. If not, use latest.
            We also don't allow repeated keys in the same file.
         """
 
```
The only variable I think we should keep constant and not change is recipe - it is implied that recipe for a build config that loads a buildspec file Is a build spec recipe. Making the variable this long isn’t as clean and nice looking :-)

shahzebsiddiqui replied:
Yea i agree, i like the word ``recipe`` its the loaded Buildspec i can revert this.

vsoch commented on buildtest/buildsystem/base.py:
```diff
@@ -457,13 +463,13 @@ def run_tests(self, testfile):
         if command.returncode == 0:
             print(
                 "{:<30} {:<30} {:<30} {:<30}".format(
```
Name is also something that I’d rather not rename like this

shahzebsiddiqui replied:
i fixed this in commit https://github.com/HPC-buildtest/buildtest-framework/pull/270/commits/e052d3d83da9ca2200bbe16ab4c6c43208b4d27c

shahzebsiddiqui committed:
applying black

Signed-off-by: Shahzeb Siddiqui <shahzebmsiddiqui@gmail.com>

vsoch reviewed (changes requested):
Good start, but let’s please limit using buildspec for only when it’s related to the file - not the loaded recipe or name for the instance.

shahzebsiddiqui committed:
use tmp_path when testing cloning of tutorials repo

shahzebsiddiqui committed:
revert some variable names as suggested by vsoch.
buildspec_recipe --> recipe
buildspec_file --> buildspec

We refer to buildspecs as one or more buildspec files while buildspec
is a single file.

shahzebsiddiqui committed:
applying black

Signed-off-by: Shahzeb Siddiqui <shahzebmsiddiqui@gmail.com>

shahzebsiddiqui committed:
change class BuildConfig to BuildspecParser

Signed-off-by: Shahzeb Siddiqui <shahzebmsiddiqui@gmail.com>

vsoch commented on buildtest/buildsystem/base.py:
```diff
@@ -1,8 +1,8 @@
 """
```
One quick thought - why not just do BuildParser?

shahzebsiddiqui replied:
I think ``BuildspecParser`` is more specific to say it's parsing and validating a ``Buildspec`` as pose to ``BuildParser``. If there are other objects other than ``Buildspec`` that buildtest may support in future then ``BuildParser`` would be too vague. It won't tell what it is parsing. Hope that makes sense

vsoch replied:
Yes, I think that's right, I'm okay with leaving the name as is, even though it's a little long :) But what other kind of Parser do you anticipate?

shahzebsiddiqui replied:
What i had in mind, though i dont have this thought out correctly is the following...

At the moment we are talking about a ``Buildspec`` that is a YAML file that is being parsed into a test script. At the moment buildtest does not have a concept of objects similar to ``kind`` in Kubernetes. In testing realm there could be objects such as ``Makefiles``, ``Dockerfile``, ``Singularityfile``.  A Buildspec that defines such an object would be parsed in a different way, or even delegated to a subclass of ``BuildspecParser`` such that it can validate the buildspec along with generated content such as Dockerfile, Singularityfile. Again these are bit out of scope and I dont have a clear picture how this will work out but certainly i think it adds value when user can specify what type of test object to generate. buildtest would need to parse correctly hence ``BuildParser`` would be bit too vague. 

There would be classes for each type of parser to ensure each artifact generated by buildtest is valid before executing.  I think it's bit hard to validate a test script or a job script without executing it. 
I think some basic validation for a test script would be something like

- check if shebang is present ``#!/bin/bash``
- check if test has read and execute permission (minimum ``555``)
- check if test has ``.sh``, ``.bash`` extension 

For jobscript it can be as simple as checking if there is atleast one scheduler directives in ``#SBATCH`` in the  file.

Again all of this is out of scope but something to think about

vsoch replied:
Oh that would make sense - if there is some other config file that warrants a special parsing. +1.

vsoch commented on buildtest/buildsystem/base.py:
```diff
@@ -208,31 +207,29 @@ class BuilderBase:
+    def __init__(self, name, recipe, buildspec=None, testdir=None):
+        """Initiate a builder base. A recipe configuration (loaded) is required.
+           this can be handled easily with the BuildspecParser class:
 
```
bc originally referred to BuildConfig so if this is now BuildParser it should be something like bp

vsoch commented on buildtest/buildsystem/base.py:
```diff
@@ -208,31 +207,29 @@ class BuilderBase:
+    def __init__(self, name, recipe, buildspec=None, testdir=None):
+        """Initiate a builder base. A recipe configuration (loaded) is required.
+           this can be handled easily with the BuildspecParser class:
 
```
bc originally referred to BuildConfig so if this is now BuildParser it should be something like bp

shahzebsiddiqui replied:
i fixed this locally.

vsoch commented on buildtest/buildsystem/base.py:
```diff
@@ -208,31 +207,29 @@ class BuilderBase:
+    def __init__(self, name, recipe, buildspec=None, testdir=None):
+        """Initiate a builder base. A recipe configuration (loaded) is required.
+           this can be handled easily with the BuildspecParser class:
 
```
bc originally referred to BuildConfig so if this is now BuildParser it should be something like bp

shahzebsiddiqui replied:
i resolved this now, changed **bc** --> **bp**

vsoch commented on buildtest/buildsystem/base.py:
```diff
@@ -208,31 +207,29 @@ class BuilderBase:
-            os.getcwd(), ".buildtest", self.config_name
-        )
+        self.buildspec = buildspec
+        self.name = re.sub("[.](yml|yaml)", "", os.path.basename(buildspec))
```
This is not correct - self.name is already set above. This is supposed to be the name of the file, not the name of the key/step.

shahzebsiddiqui replied:
I changed line 231 to use the name ``self.config_name`` i realized that was a mistake on my end, where ``self.name=name``  where name was passed as argument to ``BuilderBase``

vsoch commented on buildtest/buildsystem/base.py:
```diff
@@ -208,31 +207,29 @@ class BuilderBase:
-        )
+        self.buildspec = buildspec
+        self.name = re.sub("[.](yml|yaml)", "", os.path.basename(buildspec))
+        self.testdir = testdir or os.path.join(os.getcwd(), ".buildtest", self.name)
```
This also needs to be changed back to reference the config name.

vsoch commented on buildtest/buildsystem/base.py:
```diff
@@ -368,14 +369,14 @@ class data structures that are used by both run and dry_run
                 self.metadata[known_section] = self.recipe.get(known_section)
 
-        # Every build recipe has a shell (defaults to BUILDTEST_SHELL_DEFAULT
+        # Every Buildspec recipe has a shell (defaults to BUILDTEST_SHELL_DEFAULT)
```
Is this true?

shahzebsiddiqui replied:
hmm i guess according to schema https://github.com/HPC-buildtest/schemas/blob/master/script/script-v0.0.1.schema.json#L44 it defaults to ``bash`` there is no enum on ``shell`` so user can specify anything. 

I think the comment can be a bit misleading, every Buildspec recipe doesn't need to declare a ``shell``. It defaults to ``bash`` which is ``BUILDTEST_SHELL_DEFAULT``.

vsoch commented on buildtest/buildsystem/base.py:
```diff
@@ -457,13 +458,13 @@ def run_tests(self, testfile):
             print(
                 "{:<30} {:<30} {:<30} {:<30}".format(
-                    self.config_name, self.name, "PASSED", self.config_file
+                    self.name, self.name, "PASSED", self.buildspec
```
This needs to be changed back to be the config name, then self.name, which is the key in the file.

vsoch commented on buildtest/buildsystem/base.py:
```diff
@@ -457,13 +458,13 @@ def run_tests(self, testfile):
             print(
                 "{:<30} {:<30} {:<30} {:<30}".format(
-                    self.config_name, self.name, "FAILED", self.config_file
+                    self.name, self.name, "FAILED", self.buildspec
```
Same here....

shahzebsiddiqui replied:
This is resolved now

vsoch commented on buildtest/menu/__init__.py:
```diff
@@ -88,10 +88,9 @@ def build_menu(self):
-            "--config",
-            help="Specify test configuration file.",
-            metavar="TEST CONFIGURATION",
+            "-bs",
```
Are you sure we want to have a “bs” flag?

shahzebsiddiqui replied:
what about ``-b`` and ``--buildspec``

vsoch replied:
I like it!

shahzebsiddiqui replied:
This is resolved now in latest commit.

shahzebsiddiqui committed:
change -bs to -b as short option for --buildspec
Change self.name to self.config_name in BuilderBase class. This was
typo due to IDE Find-Replace action

vsoch commented on buildtest/menu/build.py:
```diff
@@ -10,58 +10,58 @@
 
     # If no config file provided, assume discovering across buildtest/site
-    if not config_file:
+    if not buildspec:
```
I think this config_file needs to be changed?

shahzebsiddiqui replied:
yep agreed good catch. I renamed this as ``settings_file`` since that's what we defined this based on the Terminology https://buildtest.readthedocs.io/en/devel/terminology.html.

vsoch replied:
Wait, so buildspec == settings_file ? I'm confused. If this is the settings_file, then why is it called buildspec?

vsoch commented on buildtest/menu/build.py:
```diff
@@ -143,28 +146,33 @@ def func_build_subcmd(args):
         logger.debug(f"The exclude config pattern are the following: -e {args.exclude}")
-        config_files = [
-            config for config in config_files if include_file(config, args.exclude)
+        buildspecs = [
```
If you use buildspecs here, I think we should be consistent and use the same variable in the previous function instead of buildspec_files.

shahzebsiddiqui replied:
okay i this fix this locally and call them ``buildspecs`` instead of ``buildspec_files``  this makes sense.

vsoch commented on buildtest/menu/build.py:
```diff
@@ -175,15 +183,15 @@ def func_build_subcmd(args):
-        # Read in all config files here, loading each will validate the entire file
-        bc = BuildConfig(config_file)
+        # Read in buildspec file here, loading each will validate the buildspec file
+        bc = BuildspecParser(buildspec)
```
This should be bp not bc still.

shahzebsiddiqui replied:
fixed locally. will commit soon.

vsoch commented on buildtest/settings/settings.schema.json:
```diff
@@ -1,5 +1,5 @@
 {
-  "$id": "https://raw.githubusercontent.com/HPC-buildtest/buildtest-framework/devel/buildtest/default.schema.json",
+  "$id": "https://raw.githubusercontent.com/HPC-buildtest/buildtest-framework/devel/buildtest/settings.schema.json",
```
Was this changed in the schemas repo?

shahzebsiddiqui replied:
right now this schema is in buildtest-framework as posed to schema repo. And yes this file has been updated, you can see in my branch the file is called ``settings.schema.json`` once its merged it will be in the devel location. 
https://github.com/shahzebsiddiqui/buildtest-framework/blob/regtest_fixups/buildtest/settings/settings.schema.json

vsoch replied:
Can we update the schema repository before committing changes here?

shahzebsiddiqui replied:
Recall the ``default.schema.json`` was originally part of buildtest-framework, all i did was just rename the file to comply with the correct term. Question is does it make sense to put this in schema given that only buildtest-framework will use it to validate configuration. If we put in schema repo which i don't really have any strong opinion where would we place it?

shahzebsiddiqui replied:
Also recall this schema won't be using any versions like ``script`` and other schema types. This just for validating the buildtest setting that each site will need to do to get buildtest configured.

vsoch replied:
Ah okay, then we don't need to do that.

vsoch commented on tests/buildsystem/test_base.py:
```diff
@@ -19,12 +19,12 @@ def test_load_configs():
     for config_file in os.listdir(examples_dir):
         config_file = os.path.join(examples_dir, config_file)
-        bc = BuildConfig(config_file)
+        bc = BuildspecParser(config_file)
```
bp

shahzebsiddiqui replied:
fixed locally will commit soon

vsoch commented on tests/utils/test_file.py:
```diff
@@ -1,84 +1,53 @@
+    create_dir(dir2)
+    assert is_dir(dir1)
+    assert is_dir(dir2)
+    shutil.rmtree(os.path.expandvars(os.path.join("$HOME", "a")))
```
This shutil rmtree to HOME makes me VERY uncomfortable. A test should not be trying to modify anything outside of tmp_path.

shahzebsiddiqui replied:
what would you like me to do in order to test shell and user expansion. Should i remove this test and just refactor this as part of ``resolve_path`` which just a simple assert without checking if ``create_dir`` can resolve path. The thing is ``create_dir`` doesn't use ``resolve_path`` because of the reason i mentioned in commit message https://github.com/shahzebsiddiqui/buildtest-framework/commit/2a8d6749580f2abad0c3e4c65a85b5a3d9e580e3 so i wanted to test this feature works.

shahzebsiddiqui replied:
Just to clarify these lines are needed instead of calling ``resolve_path`` since non-existent file will return ``None`` from ``resolve_path``. Since input to ``create_dir`` expects a non-existent file we can't use ``resolve_path``. We can expect ``create_dir`` to be called in this case with using ``buildtest get`` or even ``buildtest build --testdir`` where directory is auto-created. 

https://github.com/shahzebsiddiqui/buildtest-framework/blob/f8e34452871d53df2ca22b8cfb75eca1977af92b/buildtest/utils/file.py#L109-L111

i did shutil.rmtree to clean up after test creation. It may not be the best way to do it and am open to a better solution so long as we test shell and user expansion.

Without this test someone using ``buildtest build --testdir $HOME`` will result in a bug.

vsoch replied:
I think it would make sense to create an environment variable relative to the tmp_dir created by pytest, and then expand that. We should absolutely not be venturing into an actual user home and touching anything.

shahzebsiddiqui replied:
Let me see if i can use ``tmp_path`` and create a sub-directory i can only do shell expansion. User expansion is out of question since ``~`` refers to home directory. I dont have a way to test that other than what i have already.

shahzebsiddiqui replied:
okay for now i am not testing directory expansion within ``tmp_path`` I am not really sure how to go about that. 
To keep it simple i just test directory creation with a sub-directory in ``tmp_path``. See https://github.com/shahzebsiddiqui/buildtest-framework/commit/e955909d73ab4391bc2ad5864f32425e5bbf30f9

shahzebsiddiqui committed:
variable renaming as requested by @vsoch

Signed-off-by: Shahzeb Siddiqui <shahzebmsiddiqui@gmail.com>

shahzebsiddiqui committed:
rename variables, methods and comments and output where config was
referenced. This is now changed to settings or Buildspec depending
on context of sentence or purpose of variable or method.
Rename several variables defined in default.py

vsoch commented on buildtest/config.py:
```diff
@@ -33,16 +33,16 @@ def init():
-def check_configuration(config_path=None):
-    """Checks all keys in configuration file (settings/default.yml) are valid
+def check_settings(config_path=None):
+    """Checks all keys in configuration file (settings/settings.yml) are valid
```
config_path maybe changed here too?

shahzebsiddiqui replied:
yes that makes sense i will fix this in next commit

vsoch commented on buildtest/config.py:
```diff
@@ -54,34 +54,34 @@ def check_configuration(config_path=None):
-def load_configuration(config_path=None):
-    """Load the default configuration file if no argument is specified.
+def load_settings(config_path=None):
+    """Load the default settings file if no argument is specified.
```
And here as well.

shahzebsiddiqui replied:
i replaced all ``config_paths`` with ``settings_path`` in ``buildtest/config.py``. Yes there are more instance i see. Good catch

shahzebsiddiqui committed:
rename config_path with settings_path in buildtest/config.py

Signed-off-by: Shahzeb Siddiqui <shahzebmsiddiqui@gmail.com>

shahzebsiddiqui committed:
Make use of tmp_path to create temporary directory when testing
create_dir method. We create a sub-directory since tmp_path will
auto create a directory.
For now we don't test shell expansion for create_dir, it is assumed
that directory expansion test via is_dir would be sufficient for now.

vsoch reviewed (approved):
Okay, I don't want to nit forever, this looks good enough for a first go. In the future could you please try to go about small PRs? If you need to make a list of things in the description, it's probably too big. This was incredibly challenging to review - I had to look over a very large set of changes every time, and I'm not confident that I didn't miss something (e.g., a variable that was forgotten to be renamed outside of the changes that are shown).
